pkg/internal/webhook/validator.go

```diff
@@ -43,59 +41,37 @@ type validator struct {
 	registeredPlugins []string
 	webhooks          []approver.Webhook
 
-	lister  client.Reader
-	decoder *admission.Decoder
+	lister client.Reader
 }
 
-// Handle is a Kubernetes validation webhook server handler. Returns an
-// admission response containing whether the request is allowed or not.
-func (v *validator) Handle(ctx context.Context, req admission.Request) admission.Response {
-	log := v.log.WithValues("name", req.Name)
-	log.V(2).Info("received validation request")
+var _ admission.CustomValidator = &validator{}
 
-	if req.RequestKind == nil {
-		return admission.Errored(http.StatusBadRequest, errors.New("no resource kind sent in request"))
-	}
-
-	switch *req.RequestKind {
-	case metav1.GroupVersionKind{Group: policy.GroupName, Version: "v1alpha1", Kind: "CertificateRequestPolicy"}:
-		log = log.WithValues("kind", "CertificateRequestPolicy")
-
-		var policy policyapi.CertificateRequestPolicy
-		v.lock.RLock()
-		err := v.decoder.Decode(req, &policy)
-		v.lock.RUnlock()
-
-		if err != nil {
-			log.Error(err, "failed to decode CertificateRequestPolicy")
-			return admission.Errored(http.StatusBadRequest, err)
-		}
-
-		el, err := v.certificateRequestPolicy(ctx, &policy)
-		if err != nil {
-			log.Error(err, "internal error occurred validating request")
-			return admission.Errored(http.StatusInternalServerError, err)
-		}
-
-		if len(el) > 0 {
-			v.log.V(2).Info("denied admission", "errors", err)
-			return admission.Denied(el.ToAggregate().Error())
-		}
+func (v *validator) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	return v.validate(ctx, obj)
+}
 
-		log.V(2).Info("allowed request")
-		return admission.Allowed("CertificateRequestPolicy validated")
+func (v *validator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	return v.validate(ctx, newObj)
+}
 
-	default:
-		return admission.Denied(fmt.Sprintf("validation request for unrecognised resource type: %s/%s %s", req.RequestKind.Group, req.RequestKind.Version, req.RequestKind.Kind))
-	}
+func (v *validator) ValidateDelete(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	return v.validate(ctx, obj)
```

Comment:
@irbekrm why do we validate on delete?

Reply:
👍🏼 agree I've changed it in https://github.com/cert-manager/approver-policy/pull/233/commits/9bf645b2fda8506ea1478f8e188acfe98596c770